pkg/controller/common/service_control.go

```diff
@@ -53,18 +57,32 @@ func ReconcileService(
 	return reconciled, err
 }
 
+func needsDelete(expected, reconciled *corev1.Service) bool {
+	applyServerSideValues(expected, reconciled)
+	// ClusterIP is an immutable field
+	return expected.Spec.ClusterIP != reconciled.Spec.ClusterIP
+}
+
 func needsUpdate(expected *corev1.Service, reconciled *corev1.Service) bool {
-	// ClusterIP might not exist in the expected service,
-	// but might have been set after creation by k8s on the actual resource.
-	// In such case, we want to use these values for comparison.
-	if expected.Spec.ClusterIP == "" {
-		expected.Spec.ClusterIP = reconciled.Spec.ClusterIP
-	}
+	applyServerSideValues(expected, reconciled)
+	// if the specs, labels, or annotations differ, the object should be updated
+	return !(reflect.DeepEqual(expected.Spec, reconciled.Spec) &&
+		compare.LabelsAndAnnotationsAreEqual(expected.ObjectMeta, reconciled.ObjectMeta))
+}
 
+// applyServerSideValues applies any default the master might have set from the reconciled version to the expected one.
```

Comment:
`the master`?

Reply:
This is inspired by the language used in the k8s API docs

Reply:
ok 🤷‍♂ 

Still wondering if it could be worth replacing in that comment? In Elasticsearch context `master` can be confusing.
`// applyServerSideValues applies any default that may have been set from the reconciled version...` feels explicit enough and avoids ambiguity?